plugins/http_plugin/common.hpp

```diff
@@ -0,0 +1,276 @@
+            } catch(...) {}
+            return 0;
+        }
+
+        /**
+         * Helper method to calculate the "in flight" size of a std::optional<T>
+         * When the optional doesn't contain value, it will return the size of 0
+         *
+         * @param o - the std::optional<T> where T is typename
+         * @return in flight size of o
+         */
+        template<typename T>
+        static size_t in_flight_sizeof( const std::optional<T>& o ) {
+            if( o ) {
+            return in_flight_sizeof( *o );
+            }
+            return 0;
+        }
+
+        /**
+         * Helper method to calculate the "in flight" size of a string
+         * @param s - the string
+         * @return in flight size of s
+         */
+        static size_t in_flight_sizeof( const string& s ) {
+            return s.size();
+        }
+    }
+
+    // key -> priority, url_handler
+    typedef map<string,detail::internal_url_handler> url_handlers_type;
+
+    struct http_plugin_state {
+        string                         access_control_allow_origin;
+        string                         access_control_allow_headers;
+        string                         access_control_max_age;
+        bool                           access_control_allow_credentials = false;
+        size_t                         max_body_size{1024*1024};
+
+        std::atomic<size_t>                            bytes_in_flight{0};
+        std::atomic<int32_t>                           requests_in_flight{0};
+        size_t                                         max_bytes_in_flight = 0;
+        int32_t                                        max_requests_in_flight = -1;
+        fc::microseconds                               max_response_time{30*1000};
+
+        bool                     validate_host = true;
+        set<string>              valid_hosts;
+
+        bool verbose_http_errors = false;
+
+        url_handlers_type  url_handlers;
+        bool                     keep_alive = false;
+    };
+
+    /**
+     * Helper type that wraps an object of type T and records its "in flight" size to
+     * http_plugin_impl::bytes_in_flight using RAII semantics
+     *
+     * @tparam T - the contained Type
+     */
+    template<typename T>
+    struct in_flight {
+        in_flight(T&& object, http_plugin_state& plugin_state)
+        :_object(std::move(object))
+        ,_plugin_state(plugin_state)
+        {
+            _count = detail::in_flight_sizeof(_object);
+            _plugin_state.bytes_in_flight += _count;
+        }
+
+        ~in_flight() {
+            if (_count) {
+                _plugin_state.bytes_in_flight -= _count;
+            }
+        }
+
+        // No copy constructor, but allow move
+        in_flight(const in_flight&) = delete;
+        in_flight(in_flight&& from)
+        :_object(std::move(from._object))
+        ,_count(from._count)
+        ,_plugin_state(std::move(from._plugin_state))
```

Comment:
This `std::move` is not doing anything and you are copying all the plugin state here.

Reply:
plugin_state shouldn't be copied as it is a reference within in_flight object, but yes the std::move is pointless.

Reply:
plugin_state is copied. That is how references work. https://isocpp.org/wiki/faq/references#reseating-refs You can't assign references, you only copy the entire object.

Reply:
You are correct ...  I mixed up initialization with assignment.   I don't think move assignment is necessary for this code so I will delete this call.